app/src/main/java/org/thoughtcrime/securesms/conversation/v2/AttachmentDownloadHelper.kt

```diff
@@ -0,0 +1,164 @@
+                deadline = SystemClock.uptimeMillis() + BUFFER_TIMEOUT_MILLS
+                continue
+            }
+
+            if (buffer.size < BUFFER_MAX_ITEMS) {
+                // Wait for next attachment to arrive with a timeout
+                val attachment =
+                    withTimeoutOrNull((deadline - SystemClock.uptimeMillis()).coerceAtLeast(0L)) {
+                        downloadRequests.receive()
+                    }
+
+                if (attachment != null) {
+                    // Received an attachment before timeout, loop again
+                    buffer += attachment
+                    continue
+                }
+            }
+
+            // When we reach here, the buffer has reached its max size AND the timeout has expired
+            emit(buffer.toList())
+            buffer.clear()
+            deadline = -1L
+        }
+    }
+
+
+    init {
+        scope.launch {
+            downloadRequestsFlow
+                .flowOn(Dispatchers.IO)
+                .map { attachments ->
+                    val pendingAttachmentIDs = storage
+                        .getAllPendingJobs(AttachmentDownloadJob.KEY, AttachmentUploadJob.KEY)
+                        .values
+                        .mapNotNullTo(hashSetOf()) {
+                            (it as? AttachmentUploadJob)?.attachmentID
+                                ?: (it as? AttachmentDownloadJob)?.attachmentID
+                        }
+
+                    val messagesByID = mmsDatabase.getMessages(attachments.map { it.mmsId })
+                        .associateBy { it.id }
+
+                    // Before handling out attachment to the download task, we need to
+                    // check the requisite for that attachment. This check is very likely to be
+                    // performed again in the download task, but adding stuff into job system
+                    // is expensive so we need to avoid spawning new task whenever we can.
+                    attachments.filter {
+                        !pendingAttachmentIDs.contains(it.attachmentId.rowId) &&
+                                it.eligibleForDownloadTask(messagesByID[it.mmsId])
+                    }
+                }
+                .collect { attachmentsToDownload ->
+                    for (attachment in attachmentsToDownload) {
+                        jobQueue.add(
+                            AttachmentDownloadJob(
+                                attachmentID = attachment.attachmentId.rowId,
+                                databaseMessageID = attachment.mmsId
+                            )
+                        )
+                    }
+                }
+        }
+    }
+
+    private fun DatabaseAttachment.eligibleForDownloadTask(message: MessageRecord?): Boolean {
+        if (message == null) {
+            Log.w(LOG_TAG, "Message $mmsId not found for attachment $attachmentId")
+            return false
+        }
+
+        assert(message.id == mmsId) {
```

Comment:
Do we want to put anything that will fit on one line onto a single line? I'm not super fussy, but sometimes I'll put things like `if (message.isOutgoing) { return true }` rather than having it on 3 lines as long as it's still legible. Your call.